### PR TITLE
[fix](type) fix DECIMAL scale when cast function on fe

### LINKED ISSUE #12717: [Bug] Cast函数bug
### Search before asking

- [X] I had searched in the [issues](https://github.com/apache/incubator-doris/issues?q=is%3Aissue) and found no similar issues.

### Version

1.0.0-rc03

### What's Wrong?

select cast(round(16698/123,2) as DECIMAL(20,2))
预期返回135.76
实际返回135.759999999

### What You Expected?

返回135.76

### How to Reproduce?

_No response_

### Anything Else?

_No response_

### Are you willing to submit PR?

- [ ] Yes I am willing to submit a PR!

### Code of Conduct

- [X] I agree to follow this project's [Code of Conduct](https://www.apache.org/foundation/policies/conduct)

### PR DESCRIPTION
**Pull request by nextdreamblue**
# Proposed changes

Issue Number: close #xxx

## Problem summary

根据cast传递的DECIMAL类型的精度来处理DECIMAL数据.

before:
MySQL [test]> select cast('135.759999999' as DECIMAL(10,3));
+----------------------------------------+
| CAST('135.759999999' AS DECIMAL(10,3)) |
+----------------------------------------+
|                          135.759999999 |
+----------------------------------------+
1 row in set (0.00 sec)

now:
MySQL [stage]> select cast('135.759999999' as DECIMAL(10,3));
+----------------------------------------+
| CAST('135.759999999' AS DECIMAL(10,3)) |
+----------------------------------------+
|                                135.759 |
+----------------------------------------+
1 row in set (0.01 sec)

## Checklist(Required)

1. Does it affect the original behavior: 
    - [x] Yes
    - [ ] No
    - [ ] I don't know
2. Has unit tests been added:
    - [ ] Yes
    - [ ] No
    - [x] No Need
3. Has document been added or modified:
    - [ ] Yes
    - [ ] No
    - [x] No Need
4. Does it need to update dependencies:
    - [ ] Yes
    - [x] No
5. Are there any changes that cannot be rolled back:
    - [ ] Yes (If Yes, please explain WHY)
    - [x] No

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by BiteTheDDDDt** [comment]
Can you add a test case to our regression-test?

**Comment by BiteTheDDDDt** [comment]
The p1 test failed, maybe you should rebase code to pass p1.

File: regression-test/suites/tpcds_sf1000_p2/sql/q01.sql [hunk 1]
```diff
@@ -17,7 +17,7 @@ FROM
 , store
 , customer
 WHERE (ctr1.ctr_total_return > (
-      SELECT (avg(ctr_total_return) * CAST('1.2' AS DECIMAL))
```
Comment (Gabriel39) [inline regression-test/suites/tpcds_sf1000_p2/sql/q01.sql:20]: What's the difference if we still use `CAST('1.2' AS DECIMAL)`?

File: fe/fe-core/src/main/java/org/apache/doris/analysis/CastExpr.java [hunk 2]
```diff
@@ -412,7 +412,7 @@ private Expr castTo(LiteralExpr value) throws AnalysisException {
         } else if (type.isLargeIntType()) {
             return new LargeIntLiteral(value.getStringValue());
         } else if (type.isDecimalV2() || type.isDecimalV3()) {
-            return new DecimalLiteral(value.getStringValue());
```
Comment (Gabriel39) [inline fe/fe-core/src/main/java/org/apache/doris/analysis/CastExpr.java:415]: Maybe you should consider whether this is a wildcard decimal? For wildcard decimal, its scale is `-1`

File: regression-test/suites/tpcds_sf1000_p2/sql/q01.sql (hunk 1 above)
Comment (nextdreamblue) [reply regression-test/suites/tpcds_sf1000_p2/sql/q01.sql:20]: DECIMAL 默认是(9,0)吧，如果按默认走，就只剩下1了

File: fe/fe-core/src/main/java/org/apache/doris/analysis/CastExpr.java (hunk 2 above)
Comment (nextdreamblue) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/CastExpr.java:415]: 如果是-1，就和原来一样，在新的DecimalLiteral里，只有用户输入的scale大于等于0时，才去按真实的精度来处理
if (scale >= 0) {
          v = v.setScale(scale, RoundingMode.DOWN);
}

Comment (nextdreamblue) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/CastExpr.java:415]: 如果是-1，就和原来一样，在新的DecimalLiteral里，只有用户输入的scale大于等于0时，才去按真实的精度来处理
if (scale >= 0) {
v = v.setScale(scale, RoundingMode.DOWN);
}

File: regression-test/suites/tpcds_sf1000_p2/sql/q01.sql (hunk 1 above)
Comment (Gabriel39) [reply regression-test/suites/tpcds_sf1000_p2/sql/q01.sql:20]: I wonder we should treat `DECIMAL` as `DECIMAL(9, 0)` or wildcard decimal? If we get a result `1` for `CAST('1.2' AS DECIMAL)`, seems the result is wrong before?

File: fe/fe-core/src/main/java/org/apache/doris/analysis/CastExpr.java (hunk 2 above)
Comment (Gabriel39) [reply fe/fe-core/src/main/java/org/apache/doris/analysis/CastExpr.java:415]: OK. If so, I prefer
`
else if (type.isDecimalV2() || type.isDecimalV3()) {} else if (type.isWildcardDecimal())
`
and add an assertion `scale >= 0` in this new constructor.

Anyway, this is a preference and it's up to you

**Comment by hello-stephen** [comment]
https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221015072043_clickbench_pr_29646.html

**Comment by hello-stephen** [comment]
TeamCity pipeline, clickbench performance test result:
 the sum of best hot time: 35.29 seconds
 load time: 445 seconds
 storage size: 17170899412 Bytes
 https://doris-community-test-1308700295.cos.ap-hongkong.myqcloud.com/tmp/20221117111523_clickbench_pr_47918.html

**Review by Gabriel39 — APPROVED** [review]
LGTM

**Pull request merged**